### ISSUE TITLE
about vqloss

### ISSUE THREAD

**Issue by z157181773** [maincold2/Compact-3DGS#19]
Hello!
After reading your paper this year, I tried to practice reproducing your work based on the Guassian plating.
In the render function, I found that for vqloss, it doesn't seem to work? The output result is shown in the figure below. I have not learned this before, so I hope to get your help! I feel that in my code, vqloss is not playing its role
![Snipaste_2024-04-10_14-53-35](https://github.com/maincold2/Compact-3DGS/assets/122798122/0e84eab7-4e87-4aa1-8a4a-bed91f80010c)

**Comment by z157181773** [comment]
All other output data is fine

![Snipaste_2024-04-10_15-10-10](https://github.com/maincold2/Compact-3DGS/assets/122798122/0ef2ef16-330e-44bd-ba45-f6ed1ef9c336)

**Comment by maincold2** [comment]
Please refer to https://github.com/maincold2/Compact-3DGS/issues/6#issuecomment-1863688547.

Thanks.

**Comment by z157181773** [comment]
Hello! Thank you for your reply! There are still some things I don't understand that I would like to ask you for advice: I can understand that vqloss will not be used in the previous iteration, that is, codebooks will not be used. The picture above is in the number of iterations that need to use codebooks, but its value is still 0, is there no need for gradient optimization iterations? I can understand that codebook is used to replace the attributes of Gauss, reduce the storage space occupation of Gauss, do not need optimization

**Comment by z157181773** [comment]
Hello! Thank you for your reply! There are still some things I don't understand that I would like to ask you for advice: 
In your answer to https://github.com/maincold2/Compact-3DGS/issues/6#issuecomment-1863688547, I saw "we didn't use commitment loss." I don't quite understand what this means
I can understand that vqloss will not be used in the previous iteration, that is, codebooks will not be used. The picture above is in the number of iterations that need to use codebooks, but its value is still 0, is there no need for gradient optimization iterations? I can understand that codebook is used to replace the attributes of Gauss, reduce the storage space occupation of Gauss, do not need optimization

**Comment by maincold2** [comment]
Oh, you've got a nice catch!
Previous works like VQVAE or VQGAN used not only codebook loss (embedding loss) but also commitment loss.
Even without commitment loss, codebooks can be trained through codebook loss.

In the recent codebase, codebooks were optimized through EMA rather than gradient descent, which was not intended (why vqloss was 0).
And, the Pytorch VQ package we used adopts in-place optimization (learn codebook in their module), and this is also the reason for 0 vqloss.

BTW, I've now updated the code by fixing the EMA to gradient descent for in-place codebook optimization.
Also, I've verified the performances are almost the same as before.

Thanks for bringing it to my attention!
